### PR TITLE
trusted input sapling support

### PR DESCRIPTION
**Pull request by may01**
# Checklist

- [x] App update process has been followed 
- [x] Target branch is `develop` 
- [x] Application version has been bumped

### CODE REVIEW

File: src/btchip_transaction.c [hunk 1]
```diff
@@ -1104,101 +1103,379 @@ void transaction_parse(unsigned char parseMode) {
...
-                        blake2b_256_init(&btchip_context_D.transactionHashFull.blake2b, NU5_PARAM_HEADERS);
                     }
 
-                    if (btchip_context_D.transactionDataRemaining < 1) {
-                        // No more data to read, ok
-                        goto ok;
-                    }
-                    // Locktime
-                    check_transaction_available(4);
+                    // get anount of saplint spends/outputs and orchard actions
```
Comment (jmartins-fable) [inline src/btchip_transaction.c]:
```suggestion
                    // get amount of sapling spends/outputs and orchard actions

```

File: src/btchip_transaction.c [hunk 2]
```diff
@@ -1104,101 +1103,379 @@ void transaction_parse(unsigned char parseMode) {
...
-                    if (btchip_context_D.transactionDataRemaining == 0) {
+                        btchip_context_D.transactionContext.orchardActionsRemaining = btchip_context_D.orchardActionCount;
                         btchip_context_D.transactionContext.transactionState =
-                            BTCHIP_TRANSACTION_PARSED;
+                            BTCHIP_TRANSACTION_PROCESS_ORCHARD_COMPACT;
+
+                    } else if (btchip_context_D.transactionDataRemaining == 0) {
+                        btchip_context_D.transactionContext.transactionState =
+                            BTCHIP_TRANSACTION_PROCESS_EXTRA;
+                    }
```
Comment (jmartins-fable) [inline src/btchip_transaction.c:1145]: I'm having a hard time reading this portion due to the state machine logic. It is my assumption due to the standard that we can use both sapling and orchard notes during a transaction. Does this enable that? Are we cycling multiple times in here? I presume so, but just want to make sure.

File: src/btchip_transaction.c [hunk 3]
```diff
@@ -1104,101 +1103,379 @@ void transaction_parse(unsigned char parseMode) {
...
+                        // No sapling spends, just continue with extra data
+                        btchip_context_D.transactionContext.transactionState =
+                            BTCHIP_TRANSACTION_PROCESS_EXTRA; // TODO: should continue to outputs, not valid for current case
+                    }
+                    break;
+                }
+                case BTCHIP_TRANSACTION_PROCESS_SAPLING_SPENDS: {
+                    
+                    check_transaction_available(96);
+                    // update compact hash with cvg1
```
Comment (jmartins-fable) [inline src/btchip_transaction.c]: Not sure I know what this ```cvg1``` means.

File: src/btchip_transaction.c [hunk 4]
```diff
@@ -1104,101 +1103,379 @@ void transaction_parse(unsigned char parseMode) {
...
+                        
+                        blake2b_256_init(&btchip_context_D.transactionHashCompact.blake2b, (uint8_t *) NU5_PARAM_SAPLING_OUTPUTS_COMPACT);
+                        goto ok;
                     } else {
-                        btchip_context_D.transactionHashOption = 0;
-                        btchip_context_D.transactionContext.scriptRemaining =
-                            transaction_get_varint();
-                        btchip_context_D.transactionHashOption =
-                            TRANSACTION_HASH_FULL;
+                        // No sapling outputs, just continue with extra data
```
Comment (jmartins-fable) [inline src/btchip_transaction.c]: Maybe add here the why there are no sapling outputs for posteriority. We know right now, but we might not be the ones working on it next.

File: tests/test_extra_trusted_inputs.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+from ecdsa.keys import VerifyingKey
+from ecdsa.util import sigdecode_der
+
+from bitcoin_client.hwi.serialization import CTransaction
+from bitcoin_client.exception import ConditionOfUseNotSatisfiedError
+from utils import automation
+
+
+@automation("automations/accept.json")
+def test_NU5_transperent(cmd, transport):
```
Comment (joaoccmartins) [inline tests/test_extra_trusted_inputs.py]:
```suggestion
def test_NU5_transparent(cmd, transport):
```

File: tests/test_extra_trusted_inputs.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
+from hashlib import sha256
```
Comment (joaoccmartins) [inline tests/test_extra_trusted_inputs.py:1]: Can we had one that uses both Orchard and Sapling notes? AFAIK that is supported by the standard.

**Review by joaoccmartins — APPROVED** [review]
I feel like this is working, but it also seems very brittle. I don't think that's something we can do without a good refactor of the state machine logic, and I think that's out of scope in here, but I'd be remiss to say that this was an easy read. This is not on @may01 work, it's just a case of this app not being well designed to start with IMHO. 

I think we have some edge cases to take care of, but 🚀

File: tests/test_extra_trusted_inputs.py (hunk 6 above)
Comment (may01) [reply tests/test_extra_trusted_inputs.py:1]: We only fix deshielded transaction that have input from the pool and output to transparent. Such transaction's don't support input from several pools.

File: src/btchip_transaction.c (hunk 2 above)
Comment (may01) [reply src/btchip_transaction.c:1145]: We have branching here of the flow. Depending on amount of data we expect to get sapling or orchard part, we select corresponding path of parsing transaction (sapling, orchard, transparent). 
We should visit BTCHIP_TRANSACTION_OUTPUT_HASHING_DONE, one during state change second during new PADU processing.

File: Makefile [hunk 7]
```diff
@@ -28,7 +28,7 @@ APPNAME = "Zcash"
 
 # Application version
 APPVERSION_M=2
-APPVERSION_N=3
+APPVERSION_N=4
 APPVERSION_P=2
```
Comment (pvoliveira) [inline Makefile]: APPVERSION_P=0

File: src/btchip_transaction.c [hunk 8]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                    } 
+                    goto ok;
+                }
+                
+                case BTCHIP_TRANSACTION_PROCESS_SAPLING: {
+                    check_transaction_available(40);
+                    
+                    int value = 8;
+                    memcpy( btchip_context_D.saplingBalance, btchip_context_D.transactionBufferPointer, sizeof(btchip_context_D.saplingBalance));
+                    btchip_context_D.transactionBufferPointer += value;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1154]: Shouldn't we decrease `btchip_context_D.transactionDataRemaining` here?

File: src/btchip_transaction.c [hunk 9]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+
+                    
+                    value = 32;
+                    memcpy( btchip_context_D.saplingAnchor, btchip_context_D.transactionBufferPointer, sizeof(btchip_context_D.saplingAnchor));
+                    btchip_context_D.transactionBufferPointer += value;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1161]: Same here.

**Review by pvoliveira — APPROVED** [review]

File: src/btchip_transaction.c [hunk 10]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                    blake2b_256_update(&btchip_context_D.transactionHashCompact.blake2b, btchip_context_D.transactionBufferPointer, 32);
+                    btchip_context_D.transactionBufferPointer += 32;
+                    btchip_context_D.transactionDataRemaining -= 32;
+
+                    // update compact hash with rk
+                    blake2b_256_update(&btchip_context_D.transactionHashNonCompact.blake2b, btchip_context_D.transactionBufferPointer, 32);
+                    btchip_context_D.transactionBufferPointer += 32;
+                    btchip_context_D.transactionDataRemaining -= 32;
+                    
+                    btchip_context_D.transactionContext.saplingSpendRemaining -= 1;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1206]: Should we check that `saplingSpendRemaining > 0` at the beginning of this case to avoid any integer underflow?

File: tests/test_extra_trusted_inputs.py (hunk 6 above)
Comment (joaoccmartins) [reply tests/test_extra_trusted_inputs.py:1]: Probably merits a comment in the code then.

File: src/btchip_transaction.c [hunk 11]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                case BTCHIP_TRANSACTION_PROCESS_SAPLING_OUTPUTS_COMPACT: {
+
+                    long compact_size = 32+32+52; // cmu + ephemeral_key + enc_ciphertext[..52]
+                    check_transaction_available(compact_size);
+                    // update compact hash with cv
+                    blake2b_256_update(&btchip_context_D.transactionHashCompact.blake2b, btchip_context_D.transactionBufferPointer, compact_size);
+                    btchip_context_D.transactionBufferPointer += compact_size;
+                    btchip_context_D.transactionDataRemaining -= compact_size;
+                    
+                    btchip_context_D.transactionContext.saplingOutputRemaining -= 1;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1262]: Should we check that saplingSpendRemaining > 0 at the beginning of this case to avoid any integer underflow?

File: src/btchip_transaction.c [hunk 12]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                    
+                    if (btchip_context_D.transactionContext.saplingOutputRemaining ==0) {
+                        blake2b_256_init(&btchip_context_D.transactionHashMemo.blake2b, (uint8_t *) NU5_PARAM_SAPLING_OUTPUTS_MEMO);
+                        // memo_size = 512 each APDU will contain halof of the memo
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c]: I am guessing it's a quarter instead of half right?

File: src/btchip_transaction.c [hunk 13]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+
+                case BTCHIP_TRANSACTION_PROCESS_SAPLING_OUTPUTS_NONCOMPACT: {
+                    const long non_compact_size = 128; // non_compact_size = 32+16+80
+                    check_transaction_available(non_compact_size);
+                    // update compact hash with cv
+                    blake2b_256_update(&btchip_context_D.transactionHashNonCompact.blake2b, btchip_context_D.transactionBufferPointer, non_compact_size);
+                    btchip_context_D.transactionBufferPointer += non_compact_size;
+                    btchip_context_D.transactionDataRemaining -= non_compact_size;
+
+                    btchip_context_D.transactionContext.saplingOutputRemaining -= 1;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1316]: Same comment.

File: src/btchip_transaction.c [hunk 14]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                case BTCHIP_TRANSACTION_PROCESS_SAPLING_OUTPUTS_MEMO: {
+
+                    const long memo_part_size = 128; // memo_size = 512 each APDU will contain qarter of the memo
+                    check_transaction_available(memo_part_size);
+                    // update compact hash with cv
+                    blake2b_256_update(&btchip_context_D.transactionHashMemo.blake2b, btchip_context_D.transactionBufferPointer, memo_part_size);
+                    btchip_context_D.transactionBufferPointer += memo_part_size;
+                    btchip_context_D.transactionDataRemaining -= memo_part_size;
+
+                    btchip_context_D.transactionContext.saplingOutputRemaining -= 1;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1288]: Same comment.

File: src/btchip_transaction.c [hunk 15]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                }
+
+                case BTCHIP_TRANSACTION_PROCESS_ORCHARD_COMPACT: {
+                    long compact_size = 32+32+32+52; // nullifier + cmx + ephemeralKey + encCiphertext[..52]
+                    check_transaction_available(compact_size);
+                    blake2b_256_update(&btchip_context_D.transactionHashCompact.blake2b, btchip_context_D.transactionBufferPointer, compact_size);
+                    btchip_context_D.transactionBufferPointer += compact_size;
+                    btchip_context_D.transactionDataRemaining -= compact_size;
+                    
+                    btchip_context_D.transactionContext.orchardActionsRemaining -= 1;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1365]: Same comment.

File: src/btchip_transaction.c [hunk 16]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
 
+                case BTCHIP_TRANSACTION_PROCESS_ORCHARD_MEMO: {
+                    const long memo_part_size = 128; // memo_size = 512 each APDU will contain qarter of the memo
+                    check_transaction_available(memo_part_size);
+                    // update compact hash with cv
+                    blake2b_256_update(&btchip_context_D.transactionHashMemo.blake2b, btchip_context_D.transactionBufferPointer, memo_part_size);
+                    btchip_context_D.transactionBufferPointer += memo_part_size;
+                    btchip_context_D.transactionDataRemaining -= memo_part_size;
+
+                    btchip_context_D.transactionContext.orchardActionsRemaining -= 1;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1389]: Same comment

File: src/btchip_transaction.c [hunk 17]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                }
+
+                case BTCHIP_TRANSACTION_PROCESS_ORCHARD_NONCOMPACT: {
+                    const long non_compact_size = 160; // non_compact_size = 32+32+16+80
+                    check_transaction_available(non_compact_size);                    
+                    blake2b_256_update(&btchip_context_D.transactionHashNonCompact.blake2b, btchip_context_D.transactionBufferPointer, non_compact_size);
+                    btchip_context_D.transactionBufferPointer += non_compact_size;
+                    btchip_context_D.transactionDataRemaining -= non_compact_size;
+
+                    btchip_context_D.transactionContext.orchardActionsRemaining -= 1;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1414]: Same comment.

File: src/btchip_transaction.c [hunk 18]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                    blake2b_256_init(&btchip_context_D.transactionOrchardFull.blake2b, (uint8_t *) NU5_PARAM_ORCHARD);
+
+                    blake2b_256_update(&btchip_context_D.transactionOrchardFull.blake2b, orchardOutputCompactDigest, sizeof(orchardOutputCompactDigest));
+                    blake2b_256_update(&btchip_context_D.transactionOrchardFull.blake2b, orchardOutputMemoDigest, sizeof(orchardOutputMemoDigest));
+                    blake2b_256_update(&btchip_context_D.transactionOrchardFull.blake2b, orchardOutputNonCompactDigest, sizeof(orchardOutputNonCompactDigest));
+                    
+                    const long orch_dig_data_size = 1+8+32; // orchard digest data: 1+8+32
+                    check_transaction_available(orch_dig_data_size);
+                    
+                    blake2b_256_update(&btchip_context_D.transactionOrchardFull.blake2b, btchip_context_D.transactionBufferPointer, orch_dig_data_size);
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1443]: Shouldn't we update `transactionBufferPointer` and `transactionDataRemaining` here?

File: src/btchip_transaction.c [hunk 19]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
+                    btchip_context_D.transactionDataRemaining -= 4;
+                    
+                    // get extra data size
+                    btchip_context_D.transactionHashOption = 0;
+                    btchip_context_D.transactionContext.scriptRemaining = transaction_get_varint();
+                    btchip_context_D.transactionHashOption = TRANSACTION_HASH_FULL;
+                    
+                    // get expiryHeight
+                    uint8_t expiryHeight[4];
+                    memcpy(expiryHeight, btchip_context_D.transactionBufferPointer, 4);
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c]: Shouldn't we call `check_transaction_available(4);` before and then update `transactionBufferPointer` and `transactionDataRemaining`?

File: src/btchip_transaction.c [hunk 20]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
 
                     if (TX_VERSION == 5) {
-                        // We don't support sapling or orchard actions
-                        // Only expiryHeight should remain at this point
-                        if (btchip_context_D.transactionDataRemaining != 4) {
-                            PRINTF("expiryHeight expected");
-                            goto fail;
+                        check_transaction_available(4);
+                        memcpy(expiryHeight, btchip_context_D.transactionBufferPointer, 4);
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c]: Shouldn't we update transactionBufferPointer and transactionDataRemaining afterward? In this case we also ned to change the next line to:
```C
blake2b_256_update(&btchip_context_D.transactionHashFull.blake2b, expiryHeight, sizeof(expiryHeight))
```

File: src/btchip_transaction.c [hunk 21]
```diff
@@ -1104,101 +1103,377 @@ void transaction_parse(unsigned char parseMode) {
...
-                        blake2b_256_update(&btchip_context_D.transactionHashFull.blake2b, btchip_context_D.transactionBufferPointer, 4);
                     }
-
-                    transaction_offset_increase(dataAvailable);
-                    btchip_context_D.transactionContext.scriptRemaining -=
-                        dataAvailable;
+                    else {
+                        blake2b_256_update(&btchip_context_D.transactionHashFull.blake2b, locktime, sizeof(locktime));
+                        transaction_offset_increase(btchip_context_D.transactionDataRemaining);    
+                        btchip_context_D.transactionContext.scriptRemaining -= btchip_context_D.transactionDataRemaining;
```
Comment (bboilot-ledger) [inline src/btchip_transaction.c:1490]: We need to ensure that `transactionDataRemaining >= 4`

**Review by bboilot-ledger — COMMENTED** [review]
:wave: Here are the code review comments. I will do some testing tomorrow.
EDIT: testing went well, nothing more to add.

File: src/btchip_transaction.c (hunk 21 above)
Comment (may01) [reply src/btchip_transaction.c:1490]: Added check to `scriptRemaining == transactionDataRemaining` 
and decreased `scriptRemaining` before changing `transactionDataRemaining` in `transaction_offset_increase`

**Review by bboilot-ledger — APPROVED** [review]
LGTM :tada:

**Review by tdejoigny-ledger — APPROVED** [review]
